src/wallet/wallet.cpp

```diff
@@ -6126,54 +6126,134 @@ bool PaymentAddressBelongsToWallet::operator()(const libzcash::SaplingPaymentAdd
+    if (ufvkid.has_value()) {
+        // Look through the UFVKs that we have generated, and confirm that the
+        // seed fingerprint for the key we find for the ufvkid corresponds to
+        // the wallet's mnemonic seed.
+        for (const auto& [k, v] : m_wallet->mapUnifiedAccountKeys) {
+            if (v == ufvkid.value() && k.first == hdChain.value().GetSeedFingerprint()) {
+                return true;
+            }
+        }
+    }
+
     return false;
 }
 
 // GetSourceForPaymentAddress
 
-PaymentAddressSource GetSourceForPaymentAddress::operator()(const CKeyID &zaddr) const
+PaymentAddressSource GetSourceForPaymentAddress::GetUnifiedSource(const libzcash::Receiver& receiver) const
 {
-    // TODO
-    return AddressNotFound;
+    auto hdChain = m_wallet->GetMnemonicHDChain();
+    assert(hdChain.has_value());
+
+    auto ufvkMeta = m_wallet->GetUFVKMetadataForReceiver(receiver);
+    if (ufvkMeta.has_value()) {
+        auto ufvkid = ufvkMeta.value().first;
+        // Look through the UFVKs that we have generated, and confirm that the
+        // seed fingerprint for the key we find for the ufvkid corresponds to
+        // the wallet's mnemonic seed.
+        for (const auto& [k, v] : m_wallet->mapUnifiedAccountKeys) {
+            if (v == ufvkid && k.first == hdChain.value().GetSeedFingerprint()) {
+                return PaymentAddressSource::MnemonicHDSeed;
+            }
+        }
+        return PaymentAddressSource::ImportedWatchOnly;
+    } else {
+        return PaymentAddressSource::AddressNotFound;
+    }
 }
-PaymentAddressSource GetSourceForPaymentAddress::operator()(const CScriptID &zaddr) const
+
+PaymentAddressSource GetSourceForPaymentAddress::operator()(const CKeyID &addr) const
 {
-    // TODO
-    return AddressNotFound;
+    auto ufvkSource = this->GetUnifiedSource(addr);
+    if (ufvkSource == PaymentAddressSource::AddressNotFound) {
+        if (m_wallet->HaveKey(addr)) {
+            return PaymentAddressSource::Random;
+        } else {
+            if (m_wallet->HaveWatchOnly(GetScriptForDestination(addr))) {
+                return PaymentAddressSource::ImportedWatchOnly;
+            }
+        }
+    }
+
+    return ufvkSource;
+}
+PaymentAddressSource GetSourceForPaymentAddress::operator()(const CScriptID &addr) const
+{
+    auto ufvkSource = this->GetUnifiedSource(addr);
+    if (ufvkSource == PaymentAddressSource::AddressNotFound) {
+        if (m_wallet->HaveCScript(addr)) {
+            return PaymentAddressSource::Imported;
+        } else {
+            if (m_wallet->HaveWatchOnly(GetScriptForDestination(addr))) {
+                return PaymentAddressSource::ImportedWatchOnly;
+            }
+        }
+    }
+
+    return ufvkSource;
 }
 PaymentAddressSource GetSourceForPaymentAddress::operator()(const libzcash::SproutPaymentAddress &zaddr) const
 {
-    return Random;
+    if (m_wallet->HaveSproutSpendingKey(zaddr)) {
+        return PaymentAddressSource::Random;
```

Comment:
`listaddresses` does not use this payment source, instead duplicating this precise condition. Alter `listaddresses` to use this as the source of truth, as it already does for the `ImportedWatchOnly` Sprout case.

Also, I note that `listaddresses` _does_ attempt to use this visitor to look for an `Imported` Sprout case (for imported spending keys), but there is no such handler here. We should check whether we have any way to distinguish an imported Sprout spending key from a generated one, and if not, we should remove the `Imported` Sprout case from `listaddresses` (and document why it isn't checked).